working/0678 - partial classes/feature-brainstorm.md

```diff
@@ -0,0 +1,199 @@
+# Partial classes
+
+Author: kevmoo@google.com
+
+Proposed solution to [partial classes (#252)](https://github.com/dart-lang/language/issues/252).
+Discussion about this proposal should go in [Issue #678](https://github.com/dart-lang/language/issues/678).
+
+## Motivation
+
+A critical use of code generation is adding boilerplate to hand-written classes.
+
+An ideal solution would provide:
+
+- Strict separation between human- and computer-generated code.
+- Allow computer-generated code to add members (functions, fields, properties,
+  constructors, etc) to human-generated code directly – without requiring
+  subclasses, mix-ins, or manually connecting generated private members with user-created public members.
+- Allow user-created "stubs" to be added filled in by generated code. This
+  a number of static-analysis and tooling scenarios to work without the code
+  being generated first.
+
+## json_serializable before/after example
+
+Here's a motivating example with a speculative future implementation for
+[json_serializable](https://pub.dev/packages/json_serializable).
+
+### Before (current behavior)
+
+```dart
+// before: User-generated code
+import 'package:json_annotation/json_annotation.dart';
+
+part 'example.g.dart';
```

Comment:
Maybe this `part` declaration needs to be modified too, so tools won't complain if it *isn't* there yet. Currently, this file is just invalid Dart when `example.g.dart` does not exist. We likely want to continue complaining about missing imports, so it might be necessary to state that this file can be omitted temporarily for static analysis.

It is still worrysome that your code might change when importing the part file.
If the part file declares a public name, it might start shadowing an imported name, and the meaning of code changes between analysis without the part and analysis with the part.

Same can happen for identifiers inside instance members which suddenly see part-file introduced members that weren't there before.

It also means that code using the generated API still can't be analyzed before the part is available. If the `C.foo` method is only introduced by the generated code, with no stub, then any code which does `C c = ...;c.foo();` will not be correct without the generated file. You can add the stub, but then you really have to add *all* the stubs, which does not seem practical (or the goal of this approach).

Or perhaps the analyzer could have a mode where it can see that *some* import is missing *and* the library contains partial classes, so it should assume that any non-conflicting member access on that class is going to be fine (no more than a hint that 
> The C.foo member is not available yet. The following files are still missing: "example.g.dart"

or similar). It won't work if the entire class is introduced in the `part`.



The Dart system patch files are similar in some ways and they cannot introduce new public names, and the library itself cannot see private names from the patch file (it can define an `external` private member first, then the patch can provide the implementation). This also means that the generated code can use any private name for its implementation without risking conflicts.

That approach conflicts with goal of allowing generated code to add public members which do not have stubs, though.